cpp/include/raft/neighbors/ivf_flat_types.hpp

```diff
@@ -76,6 +76,12 @@ struct index_params : ann::index_params {
    * flag to `true` if you prefer to use as little GPU memory for the database as possible.
    */
   bool conservative_memory_allocation = false;
+  /**
+   * Seed used for random sampling if kmeans_trainset_fraction < 1.
+   *
+   * Value -1 disables random sampling, and results in sampling with a fixed stride.
+   */
+  int random_seed = 0;
```

Comment:
More and more algorithms in raft depend on the random generator.
In IVF methods in particular, we are going to have two random states/generators for the subsampling and kmeans clustering.
Maybe it's a good time we introduce a new `raft::resource` `RNG_STATE` and remove the seed parameter from the explicit algo parameters everywhere? I see following advantages of this:
  - The random state shared across all algorithms removes unwanted possible correlation coming from using the same seed (especially when the same seed is passed to a repeatedly invoked inner algorithm, such as kmeans clustering).
  - No need to re-export the `seed` parameter of the inner algorithm to the outer algorithm.
  - A centralized place to set random seed (e.g. `raft::resource::set_rng_seed(res, seed)`)

CC @cjnolet

Reply:
I like the idea.

To clarify: we would have a single random sub-sampling. The sub-sampling for codebooks (#2052) would just take the first n points from the already randomized elements.

The `random_seed` variable is currently also used to switch back to fixed stride sampling. Do we want to keep such switch, or shall we just remove fixed stride sampling altogether?

Reply:
I was actually thinking about the random initialization in [kmeans_balanced::build_clusters](https://github.com/rapidsai/raft/blob/fafb8ed6c56e8c2ff08d6174fb9b6a4c65fac9cf/cpp/include/raft/cluster/detail/kmeans_balanced.cuh#L721-L726) where the random seed would be useful as well, but apparently it's not so random so far :)

I'd prefer to scrap the fixed stride sampling altogether if the performance impact is not a blocker. Less code is always better!

Reply:
Capturing our offline discussion:
- We did not reach agreement yet on whether to add `random_seed` to `raft::resources`.
- In order to avoid breaking API changes, we can start with a fixed seed, and update later.
- If the perf is not affected, then we can remove fixed stride subsampling.

Reply:
Closing this thread, let's continue on #2105.